pkg/istio-agent/health/health_probers.go

```diff
@@ -54,6 +58,49 @@ func (p *ProbeResult) IsHealthy() bool {
 	return *p == Healthy
 }
 
+type GRPCProber struct {
+	Config *v1alpha3.GrpcHealthCheckConfig
+}
+
+var _ Prober = &GRPCProber{}
+
+func NewGRPCProber(cfg *v1alpha3.GrpcHealthCheckConfig) *GRPCProber {
+	return &GRPCProber{
+		Config: cfg,
+	}
+}
+
+func (g *GRPCProber) Probe(timeout time.Duration) (ProbeResult, error) {
+	if g.Config == nil {
+		return Unknown, fmt.Errorf("grpc health check config is nil")
+	}
+	conn, err := grpc.NewClient(net.JoinHostPort("localhost", fmt.Sprintf("%d", g.Config.Port)),
+		grpc.WithUserAgent("istio-probe/1.0"),
+		grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return Unknown, fmt.Errorf("failed to create connection on port %d for grpc health check: %v", g.Config.Port, err)
+	}
+	defer conn.Close()
+	client := grpc_health_v1.NewHealthClient(conn)
+	req := &grpc_health_v1.HealthCheckRequest{
+		Service: g.Config.Service,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
```

Comment:
the timeout should include the dial time, right now it ignores it and only starts the timer after the connection is open